core/search/conjunction.hpp

```diff
@@ -79,182 +87,360 @@ struct score_iterator_adapter {
     }
+  }
 
-    // prepare score
-    switch (scores_.size()) {
-      case 0:
-        IRS_ASSERT(score.IsDefault());
-        score = ScoreFunction::Default(Merger::size());
-        break;
-      case 1:
-        score = std::move(*scores_.front());
-        break;
-      case 2:
-        score.Reset(*this, [](score_ctx* ctx, score_t* res) noexcept {
-          auto& self = *static_cast<conjunction*>(ctx);
-          auto& merger = static_cast<Merger&>(self);
-          (*self.scores_.front())(res);
-          (*self.scores_.back())(merger.temp());
-          merger(res, merger.temp());
-        });
-        break;
-      case 3:
-        score.Reset(*this, [](score_ctx* ctx, score_t* res) noexcept {
-          auto& self = *static_cast<conjunction*>(ctx);
-          auto& merger = static_cast<Merger&>(self);
-          (*self.scores_.front())(res);
-          (*self.scores_[1])(merger.temp());
-          merger(res, merger.temp());
-          (*self.scores_.back())(merger.temp());
-          merger(res, merger.temp());
-        });
-        break;
-      default:
-        score.Reset(*this, [](score_ctx* ctx, score_t* res) noexcept {
-          auto& self = *static_cast<conjunction*>(ctx);
-          auto& merger = static_cast<Merger&>(self);
-          auto begin = std::begin(self.scores_);
-          auto end = std::end(self.scores_);
-
-          (**begin)(res);
-          for (++begin; begin != end; ++begin) {
-            (**begin)(merger.temp());
-            merger(res, merger.temp());
-          }
-        });
-        break;
+  attribute* get_mutable(irs::type_info::type_id type) noexcept final {
+    return irs::get_mutable(attrs_, type);
+  }
+
+  IRS_FORCE_INLINE auto& score() { return std::get<irs::score>(attrs_).max; }
+
+  doc_id_t value() const final { return std::get<document>(attrs_).value; }
+
+  bool next() override { return !doc_limits::eof(seek(value() + 1)); }
+
+  doc_id_t shallow_seek(doc_id_t target) final {
+    target = ShallowSeekImpl(target);
+    if (IRS_UNLIKELY(doc_limits::eof(target))) {
+      return Seal();
     }
+    return leafs_doc_;
   }
 
-  // tries to converge front_ and other iterators to the specified target.
-  // if it impossible tries to find first convergence place
-  doc_id_t converge(doc_id_t target) {
-    IRS_ASSERT(!doc_limits::eof(target));
+  doc_id_t seek(doc_id_t target) override {
+    auto& doc = std::get<document>(attrs_).value;
+    if (IRS_UNLIKELY(target <= doc)) {
+      if constexpr (Root) {
+        if (threshold_ < score_) {
+          return doc;
+        }
+        target = doc + !doc_limits::eof(doc);
+      } else {
+        return doc;
+      }
+    }
+    auto& merger = static_cast<Merger&>(*this);
+  align_leafs:
+    target = ShallowSeekImpl(target);
+  align_docs:
+    if (IRS_UNLIKELY(doc_limits::eof(target))) {
+      return Seal();
+    }
+    auto it = this->itrs_.begin();
+    const auto seek_target = (*it)->seek(target);
```

Comment:
Suggest to remove `seek_target` and just always store sekk result in `target`

Reply:
Well, it's not possible check line 315

Reply:
True.